### ISSUE TITLE
Use React 16's forwardRef for most/all components

### ISSUE THREAD

**Issue by seansfkelley** [reactstrap/reactstrap#1082]
reactstrap is a perfect use-case for React 16's new [`forwardRef` API](https://reactjs.org/blog/2018/03/29/react-v-16-3.html#forwardref-api). Since most components in this library have predictable and well-defined DOM they produce, it would be very useful to have them forward their refs along to the underlying elements so consumers like myself don't have to wrap access in `ReactDOM.findDOMNode` when we want to e.g. implement a scroll-related feature.

**Comment by juanmaguitar** [comment]
Totally Agree!! We're using some `reactstrap` components in our project and in order to expand the current behavior it would be really handy (and much more consistent) being able to select the DOM elements via `ref` selections in the same components

For example, doing this in the `Tooltip` component 

```
class Tooltip extends React.Component {
  constructor(props) {
    super(props);

    this.refTooltipInner = React.createRef();
    ...
  }

  render() {
    ...
     return (
      <PopperContent
        ...
      >
        <div
          ref={this.refTooltipInner}
          ...
        />
      </PopperContent>
    );
  }
}
```

Would allow to have access to this DOM element easily _from the outside_

**Comment by TheSharpieOne** [comment]
I'm all for this, but it will be a breaking change since it will change the require version of react needed as a peerDep.

**Comment by illiteratewriter** [comment]
If adding `refs` is all that is needed, we can do that like how we did [https://github.com/reactstrap/reactstrap/issues/1054](url), so that we don't have to change react version, right? So wouldn't it be better to do it like that so we don't have to bumb the version up?

**Comment by seansfkelley** [comment]
True. You could put `innerRef` on everything, and use this issue as a tracking issue for the eventual deprecation and replacement with `forwardRef` as noted in #1054.

**Comment by seansfkelley** [comment]
@TheSharpieOne when would you accept a PR for this breaking change? I keep coming up against this cause I want to attach refs to things that often turn out to be stateless components. I'd be open to making a PR to apply the change as well. We might be able to keep `innerRef` around as a deprecated alias to preserve API-compatibility even as the minimum React version is raised.

It's a shame that 16.3 was introduced not 6 days after #916 dropped support for versions below 16. :(

**Comment by TheSharpieOne** [comment]
there is a v7 branch, make PRs with breaking changes to that branch, eventually it will be merged and released.

**Comment by seansfkelley** [comment]
@TheSharpieOne looks like I just missed the chance to get into v7, but I just opened #1356 for the nest release.

**Comment by iamandrewluca** [comment]
> When you start using forwardRef in a component library, you should treat it as a breaking change and release a new major version of your library. This is because your library likely has an observably different behavior (such as what refs get assigned to, and what types are exported), and this can break apps and other libraries that depend on the old behavior.
> https://reactjs.org/docs/forwarding-refs.html#note-for-component-library-maintainers

**Comment by TheSharpieOne** [comment]
Yes, it would be nice to get forwardRef into the next major release. This missed v7 as well as v8

**Comment by iamandrewluca** [comment]
@TheSharpieOne Now I'm working on https://github.com/reactjs/react-transition-group/pull/559 to migrate from deprecated `findDOMNode` then I think I'll create a new PR from #1356, to migrate to forward ref, and use new `react-transition-group` api, so we can finally use `React.StrictMode` in our app.

**Comment by darron1217** [comment]
Any updates on this issue?
It is really needed.
